ood-portal-generator/lib/ood_portal_generator/application.rb

```diff
@@ -29,41 +33,214 @@ def context
+        checksum = checksum_str.split(' ')[0]
+
+        str = parse_input(input)
+        new_checksum = Digest::SHA256.hexdigest(str)
+
+        checksum == new_checksum
+      end
+
+      def checksum_exists?
+        return false unless File.exist?(sum)
+        File.readlines(sum).grep(apache).size == 0
+      end
+
+      def update_replace?
+        # If checksum of ood-portal.conf matches or ood-portal.conf doesn't exit, replace is possible.
+        # If the checksum matches this means a site has not changed ood-portal.conf outside ood-portal-generator
+        if checksum_matches?(apache)
+          replace = true
+        else
+          replace = false
+        end
+        replace = true if force
+        return replace
+      end
+
+      def exit!(code)
+        case code
+        when TrueClass
+          exit(0)
+        when FalseClass
+          exit(1)
+        else
+          exit(code)
+        end
+      end
+
+      def update_ood_portal
+        ret = 0
+        changed = false
+        new_apache = Tempfile.new('new_apache')
+        @output = new_apache.path
+        self.start('generate')
+
+        # Create checksum file if the path to ood-portal.conf not in checksum file
+        # Checksum is based on mktemp generated ood-portal.conf but using path of real ood-portal.conf
+        if ! checksum_exists?
+          puts "Generating Apache config checksum file: '#{sum}'"
+          save_checksum(new_apache.path)
+        end
+
+        replace = update_replace?
+
+        if ! FileUtils.cmp(new_apache.path, apache)
+          if replace
+            if File.exist?(apache)
+              puts "Backing up previous Apache config to: '#{apache_bak}'"
+              FileUtils.mv(apache, apache_bak)
+            end
+            puts "Generating new Apache config at: '#{apache}'"
+            `cat "#{new_apache.path}" > "#{apache}"`
+            puts "Generating Apache config checksum file: '#{sum}'"
+            save_checksum(apache)
+            ret = change_exit
+            changed = true
+          else
+            puts "WARNING: Checksum of #{apache} does not match previous value, not replacing."
+            puts "Generating new Apache config at: '#{apache}.new'"
+            `cat "#{new_apache.path}" > "#{apache}.new"`
+            ret = skip_exit
+          end
+        else
+          puts "No change in Apache config."
+        end
+
+        new_apache.unlink
+
+        if rpm || ! replace
+          return ret
+        end
+
+        puts "Completed successfully!"
+        if changed
+          puts ""
+          puts "Restart the #{apache_services[0]} service now."
+          puts ""
+          puts "Suggested command:"
+          puts "    sudo systemctl try-restart #{apache_services[0]}.service #{apache_services[1]}.service"
+          puts ""
+        end
+
+        return ret
+      end
+
       # Starts the OodPortalGenerator CLI
       # @return [void]
-      def start
+      def start(mode)
```

Comment:
Instead of having start(mode) and then parser(mode) (control coupling in two places) it might be preferable to just have two separate methods. See bottom of the comment.

On the other hand, update_ood_portal does execute generate which runs through the optparser, so perhaps for update_ood_portal all of the CLI arguments are valid including output intput and config. Finally, I remember seeing this for using OptionParser with subcommands - it is a variation on what you are doing here https://gist.github.com/rkumar/445735. I don't know if it is worse or better but I was thinking about using something similar for https://github.com/OSC/ondemand/issues/15.

```ruby
def cli_generate(argv=ARGV)
  Process.setproctitle("generate  #{argv.join(" ")}")

  OptionParser.new do |parser|
    # generator options

    add_shared_opt_parser_attrs parser
  end.parse! argv

  view = View.new(context)
  rendered_template = view.render(template.read)
  output.write(rendered_template)
end

def cli_update_ood_portal
  Process.setproctitle("generate  #{argv.join(" ")}")

  OptionParser.new do |parser|
    # update_ood_portal options

    add_shared_opt_parser_attrs parser
  end.parse! argv

  exitcode = update_ood_portal()
  exit!(exitcode)
end

def add_shared_opt_parser_attrs(parser)
  parser.on("-v", "--version", "Print current version") do
    puts "version #{OodPortalGenerator::VERSION}"
    exit
  end
  parser.on("-h", "--help", "Show this help message") do
    puts parser
    exit
  end
  parser.separator ""
  parser.separator "Default:"
  parser.separator "  generate -c #{config} -t #{template}"
end
```

Reply:
Something similar to this was implemented